scripts/commands/checkPatternsIndex.ts

```diff
@@ -0,0 +1,212 @@
+// This code is a Qiskit project.
+//
+// (C) Copyright IBM 2024.
+//
+// This code is licensed under the Apache License, Version 2.0. You may
+// obtain a copy of this license in the LICENSE file in the root directory
+// of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+//
+// Any modifications or derivative works of this code must retain this
+// copyright notice, and modified files need to carry a notice indicating
+// that they have been altered from the originals.
+
+import { readFile } from "fs/promises";
+import { pathExists } from "../lib/fs";
+
+const IGNORE_URL = ["/guides/qiskit-serverless"];
+
+const INDEX_PAGES = [
+  "docs/guides/map-problem-to-circuits.mdx",
+  "docs/guides/optimize-for-hardware.mdx",
+  "docs/guides/execute-on-hardware.mdx",
+  "docs/guides/postprocess-results.mdx",
+];
+
+const TOC_PATH = `docs/guides/_toc.json`;
+
+async function getIndexEntries(indexPage: string): Promise<string[]> {
+  const rawIndex = await readFile(indexPage, "utf-8");
+  const result: string[] = [];
+  for (const line of rawIndex.split("\n")) {
+    // The index represents an unordered list of entries starting with `*`.
+    if (!line.trimStart().startsWith("* ")) {
+      continue;
+    }
+
+    const slug = extractPageSlug(line);
+    if (slug) {
+      result.push(slug);
+    }
+  }
+
+  return result;
+}
+
+function extractPageSlug(text: string): string | undefined {
+  const re = /\((.*)\)/gm;
+  // Ex: '* [Circuit library](./circuit-library)'.
+  const match = re.exec(text);
+  if (!match) {
+    // Nested sections don't have any link
+    return undefined;
+  }
+  const pageSlug = match[1];
+  if (pageSlug.startsWith("http") || pageSlug.startsWith("/")) {
+    return pageSlug;
+  }
+  return `/guides/${pageSlug.split("/").pop()}`;
+}
+
+function getTocSectionPageNames(sectionNode: any): string[] {
+  let results = [];
+  if (sectionNode.url) {
+    results.push(sectionNode.url);
+  }
+
+  if (sectionNode.children) {
+    for (const child of sectionNode.children) {
+      results = [...results, ...getTocSectionPageNames(child)];
+    }
+  }
+
+  return results;
+}
+
+async function getToolsTocEntriesToCheck(): Promise<string[]> {
+  const toc = JSON.parse(await readFile(TOC_PATH, "utf-8"));
+  const toolsNode = toc.children.find((child: any) => child.title == "Tools");
+  const toolsPages = getTocSectionPageNames(toolsNode);
+  return toolsPages.filter((page) => !IGNORE_URL.includes(page));
+}
+
+async function getDeduplicateEntriesAndAddErrors(
```

Comment:
A couple suggestions on this function:

1. It'd be better to move the `const entries =` back into `main`, imo. This function is doing a lot right now, which you can tell by the function name. Instead, this function would take `entries: string[]` as an arg
2. The function should be returning a `Set<string>` for the deduplicated entries since it has O(1) lookup
3. You can call the function something like `deduplicateEntries`
4. Rather than mutating the `errors` arg, more idiomatic is to return the errors as a new value. I know that you call this function twice; you'd concatenate the two error arrays in the final place that you use them. You'd end up returning `[Set<string>, string[]]`, or you could define an object to be more explicit.

Reply:
Changed it to follow your comment :+1: